plugins/telink-dfu/fu-telink-dfu-hid-device.c

```diff
@@ -0,0 +1,370 @@
+{
+#if defined(HAVE_HIDRAW_H) && FU_TELINK_DFU_HID_DEVICE_SHOW_DESC_INFO
+	FuHidDevice *hid_dev = FU_HID_DEVICE(self);
+	g_autoptr(GPtrArray) descriptors = NULL;
+	g_autoptr(GString) desc_err = g_string_new(NULL);
+
+	descriptors = fu_hid_device_parse_descriptors(hid_dev, error);
+	if (descriptors == NULL)
+		return;
+	g_debug("descriptors=%u", descriptors->len);
+	for (guint i = 0; i < descriptors->len; i++) {
+		g_autoptr(GError) err = NULL;
+		if (desc_err->len > 0) {
+			g_string_append(desc_err, ", ");
+		}
+		g_string_append_printf(desc_err,
+				       "descriptor 0x%x: %s",
+				       i,
+				       err->message);
+	}
+#endif
+}
+
+static gboolean
+fu_telink_dfu_hid_device_setup(FuDevice *device, GError **error)
+{
+	FuTelinkDfuHidDevice *self = FU_TELINK_DFU_HID_DEVICE(device);
+
+	/* HidDevice->setup */
+	if (!FU_DEVICE_CLASS(fu_telink_dfu_hid_device_parent_class)->setup(device, error))
+		return FALSE;
+
+	fu_device_set_name(device, "telink-hid-device");
+
+	fu_telink_dfu_hid_device_show_descriptors(self, error);
+	fu_telink_dfu_hid_device_show_edp_info(self, error);
+
+	/* success */
+	return TRUE;
+}
+
+static FuStructTelinkDfuHidPkt *
+fu_telink_dfu_hid_device_create_packet(FuTelinkDfuCmd cmd, const guint8 *buf, gsize bufsz, GError **error)
+{
+	FuStructTelinkDfuHidPkt *pkt = fu_struct_telink_dfu_hid_pkt_new();
+	guint8 preamble[6] = {
+		FU_TELINK_DFU_HID_DEVICE_REPORT_ID,
+		FU_TELINK_DFU_HID_DEVICE_OP_CODE,
+		20,
+		0x00,
+		(guint8)(cmd & 0x00ff),
+		(guint8)(cmd >> 8),
+	};
+	guint8 payload_len = 0;
+
+	switch (cmd) {
+	case FU_TELINK_DFU_CMD_OTA_FW_VERSION:
+		payload_len = 0;
+		break;
+	case FU_TELINK_DFU_CMD_OTA_START:
+		payload_len = 2;
+		break;
+	case FU_TELINK_DFU_CMD_OTA_END:
+		payload_len = 6;
+		break;
+	default:
+		/* OTA data */
+		payload_len = 20;
+		break;
+	}
+	preamble[2] = payload_len;
+
+	fu_struct_telink_dfu_hid_pkt_set_preamble(pkt, preamble, sizeof(preamble), error);
+	if (buf != NULL)
+		fu_struct_telink_dfu_hid_pkt_set_payload(pkt, buf, bufsz, NULL);
+	/* payload starts from preamble[4]; the cmd field is excluded from crc calcualtion */
+	fu_struct_telink_dfu_hid_pkt_set_crc(pkt, ~fu_crc16(FU_CRC_KIND_B16_USB, pkt->data + 4, pkt->len - 6));
+	return pkt;
+}
+
+static gboolean
+fu_telink_dfu_hid_device_write(FuTelinkDfuHidDevice *self,
+			       guint8 *buf,
+			       gsize bufsz,
+			       GError **error)
+{
+#ifdef HAVE_HIDRAW_H
+	FuUsbDevice *usb_dev = FU_USB_DEVICE(FU_DEVICE(self));
+	FuHidDevice *hid_dev = FU_HID_DEVICE(self);
+	g_autoptr(GPtrArray) ifaces = NULL;
+	guint8 ota_buf[FU_TELINK_DFU_HID_DEVICE_OTA_LENGTH] = {0};
+
+	memcpy(ota_buf, buf, MIN(FU_TELINK_DFU_HID_DEVICE_OTA_LENGTH, bufsz));
+
+	ifaces = fu_usb_device_get_interfaces(usb_dev, error);
+	//the last interface would always be reserved for OTA upgrade 
+	fu_hid_device_set_interface(hid_dev, ifaces->len - 1);
+	fu_hid_device_set_ep_addr_in(hid_dev, HID_EP_IN);
+	fu_hid_device_set_ep_addr_out(hid_dev, HID_EP_OUT);
```

Comment:
i think it's a lot safer to do this in ->probe()

Reply:
resolved in commit 7857a6cb0d8cf1922ec1aef16606880f73440f58